### PR TITLE
Milestone5: Friendships v1 🚀

### PR DESCRIPTION
**Pull request by KingsleyMcSimon**
In this pull request, we  worked in friendship in active record, First Create a FriendShip table, Amend the belong to Relationships in the Friendship model, Create has many relationships in the user model, create a friends methods, pendings_friends, confirm_friends in the user model.
Milestone 5: Friendships v1

### CODE REVIEW

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
#### Hello Team 👋
#### Status: Changes Required

Great job so far! 👏👋
The associations seem to be correct.

> This is a social media app where we supposed to build a web application with friendship architecture. A user should be able to send friend requests and accept or reject the requests. I noted down all the requirements that you need to fulfill one by one. Please carefully read them and try to implement them.

> Also, please read [this](https://smartfunnycool.com/friendships-in-activerecord/) article to have a good understanding of the friendship architecture in Rails.

Let's cover the details together to make your project better 😊⏬

### GitHub Actions

- [x] [REQUIRED] Please make sure that your GitHub Actions feature is enabled by checking repository settings.

### Application Behavior

- [x] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  Currently, there is no invitation feature.

- [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  Currently, I can't send friend requests to other users.

- [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  Currently, I can't send friend requests to other users.

- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  Currently, I can't send friend requests to other users.

- [x] There is a button to invite the user to friendship on users list page and on single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  Currently, there is no invitation button for friendship.

- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

:bulb:  After implying friendship architecture, please work on this too.

### Testing

- [x] After implementing all of these features, please also add the necessary tests for your application.

### Views & Helpers

- [x] There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers.
- [x] View-related logic can be encapsulated in View helpers (app/helpers/\*\_helper.rb)
      **Resources**
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

:bulb:  Currently, there is "logic" in the `application.html.erb` file. I know it is coming from the forked repository. But let's make it correct by refactoring our code.

## That's it, you can request another code review request after making the changes ♻️😇

With all my respect 🙏
> Kubilay Caglayan | Technical Support Engineer

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello Team
### Good job so far :clap: However, you did not make all the changes from the previous review. 
### Make all the changes especially the app behaviour and submit for another code review. Good luck.
- Happy coding!

**Comment by JasemDuncan** [comment]
Thanks @rOluochKe, please can u give us a hint. We want to implement all the requested changes.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

### However, we need to work on some changes before we can merge.

- [x] Please enable GitHub actions from your settings. 

- [x] Please add a button to invite the user to friendship on the single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- While "sending" and "rejecting" friendships work, "accepting" a friendship request throws this error:

  image

  - [x] Please investigate this and fix it.

- **Reminder**  The "timeline" will only consist of posts from the logged-in user and their friends. Just check this functionality once you fix the above error.

- Please add tests for this milestone. 
   
  - [x] Add tests for **all** of the models, not just the ones you created.

  - [x] Add tests for all the controllers, not just the ones you created. RSpec recommends using request specs to test controllers but it's completely up to you.

  - [x] Add tests for all the features, not just the ones you created.
  
  **IMPORTANT!** All of your tests should cover both positive and negative scenarios.

- There is still some logic left in `app/views/layouts/application.html.erb`. I know it is coming from the forked repository. But let's make it correct by refactoring our code.

  - [x] Please move this logic to `/helpers`.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍impressive work on making the changes! you still need some fixes to do 🛠️

## Required Changes ♻️ 

- [ ] You still haven't covered all tests as requested by the previous TSEs

- [x] Enable GitHub actions on this Pull Request, learn more [here](https://github.com/microverseinc/linters-config/tree/master/ror)

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

File: app/models/friendship.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Friendship < ApplicationRecord
+  belongs_to :user
+  belongs_to :friend, class_name: 'User'
+end
```
Comment (ghost) [inline app/models/friendship.rb:4]: Write tests for the above associations.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team 😃,

**Great job implementing the requested changes 💪. Here are a few fixes that are still required.**

1). **Your feature specs are failing.**
spec

2). **Add the following.**

- **feature spec tests for user authentication, (they are added in `spec/feature` folder).**

- **feature spec tests for posts,  (they are added in `spec/feature` folder).**

- **spec tests for your helpers and controllers.**

3). **Your GitHub actions on this pull request is not enabled. Hint: You can go to settings -> actions. Under Actions permissions, choose disable actions option and click save. Then refresh the same page, choose allow all actions option and click save. If that still won't work for you, then please reach out to one of the full-time TSEs for more guidance.**

4). **There are still some rubocop offenses.**
rubocop

5). **See one more comment below.**

### Happy coding 😃

**PS: You can reach me on slack (Rose Wanjohi).**

**Comment by Dandush03** [comment]
# Status: :rocket: Approved!
## Great Work on all your implementation
# Happy Coding :nerd_face:

**Pull request merged**